File: Day34/subarraySum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function (nums, k) {

    let map = {
        "0": 1
    }

    let sum = 0;
    let total = 0;

    for (let i = 0; i < nums.length; i++) {
        sum += nums[i];

        if ((sum - k) in map) {
            total += map[sum - k];
        }

        map[sum] = (map[sum] || 0) + 1;

    }

    return total;
};

File: Day34/subarraySum.cpp
class Solution
{
public:
    int subarraySum(vector<int> &nums, int k)
    {
        std::unordered_map<int, int> map{{0, 1}};
        int total = 0;
        int sum = 0;

        for (int i = 0; i < nums.size(); i++)
        {
            sum += nums[i];

            if (map.count(sum - k))
            {
                total += map[sum - k];
            }

            map[sum]++;
        }

        return total;
    }
};